### PR TITLE
[BUGFIX] refactor inheritance process

### PR DESCRIPTION
**Pull request by pixelmatseriks**
# New Pull Request checklist

## PR Type

What kind of change does this PR introduce?

- [ ] [FEATURE] - A new feature
- [x] [BUGFIX] - A bug fix
- [ ] [TASK] - Any task, which is not a **new feature** or **bugfix**
- [ ] [TEST] - Adding missing tests
- [ ] [DOC] - Documentation only changes
- [ ] [WIP] - Work in progress tag, should not be present when creating pull requests

## Breaking change

Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: Classes/Utility/DataInheritanceUtility.php [hunk 1]
```diff
@@ -37,8 +57,806 @@ public static function getInheritedFieldsForProductType(int $productType): array
...
+        $identifierField = $eventDispatcher
+            ->dispatch(new InlineIdentifierFieldEvent($foreignTable))
+            ->getField();
+
+        foreach ($resolvedRelationsItems as $index => $resolvedItem) {
+            $identifier = BackendUtility::getRecord(
+                $resolvedItem['table'],
+                $resolvedItem['uid'],
+                $identifierField
+            )[$identifierField];
```
Comment (mabolek) [inline Classes/Utility/DataInheritanceUtility.php:602]: What do you do when BackendUtility::getRecord() returns null?

File: Classes/Utility/DataInheritanceUtility.php [hunk 2]
```diff
@@ -14,6 +30,10 @@ class DataInheritanceUtility
 {
```
Comment (mabolek) [inline Classes/Utility/DataInheritanceUtility.php:30]: Some really great stuff in this class!!!

File: Configuration/Services.yaml [hunk 3]
```diff
@@ -17,12 +17,60 @@ services:
...
+      - name: event.listener
+        identifier: 'calculateInlineFieldAttributeValueValueEvent'
+        before: ''
+        event: Pixelant\PxaProductManager\Event\DataInheritance\CalculateInlineFieldValueEvent
+
+  Pixelant\PxaProductManager\Event\DataInheritance\InheritNewInlineFileReferenceDataEvent:
+    tags:
+      - name: event.listener
+        identifier: 'inheritNewInlineFileReferenceDataEvent'
+        before: ''
```
Comment (mabolek) [inline Configuration/Services.yaml:50]: Isn't it possible to just skip the `before`?

File: ext_localconf.php [hunk 4]
```diff
@@ -121,8 +121,9 @@
 
 
     // Modify data structure of flexform. Hook will dynamically load flexform parts for selected action
-    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS'][\TYPO3\CMS\Core\Configuration\FlexForm\FlexFormTools::class]['flexParsing']['pxa_product_manager'] =
-        \Pixelant\PxaProductManager\Hook\FlexFormDataStructureHook::class;
+    // Disable hook, problems during translations import. Don't think it is needed anymore.
+    // $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS'][\TYPO3\CMS\Core\Configuration\FlexForm\FlexFormTools::class]['flexParsing']['pxa_product_manager'] =
+    //    \Pixelant\PxaProductManager\Hook\FlexFormDataStructureHook::class;
```
Comment (mabolek) [inline ext_localconf.php:126]:
```suggestion
```

**Review by mabolek — CHANGES_REQUESTED** [review]

File: Classes/Command/FixDuplicateAttributeValuesCommand.php [hunk 5]
```diff
@@ -93,94 +136,257 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+    }
+
+    /**
+     * Fetch attribute values by product and attribute.
+     *
+     * @param int $product
+     * @param int $attribute
+     * @return array
+     */
+    protected function fetchAttributeValuesByProductAndAttribute(int $product, int $attribute): array
```
[the remark this answers is not in the document]
Comment (pixelmatseriks) [reply Classes/Command/FixDuplicateAttributeValuesCommand.php:195]: Hmm, I'm not sure, this function returns all attributevalues by product and attribute, hopefully it won't be needed in future.

File: Classes/Command/FixDuplicateAttributeValuesCommand.php [hunk 6]
```diff
@@ -93,94 +136,257 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        return $records;
+    }
+
+    /**
+     * Fetch attribute value by id.
+     *
+     * @param int $id
+     * @return array
+     */
+    protected function fetchAttributeValueById(int $id): array
```
[the remark this answers is not in the document]
Comment (pixelmatseriks) [reply Classes/Command/FixDuplicateAttributeValuesCommand.php:227]: Not same table, attributevalue vs attribute.

File: Classes/Command/FixDuplicateAttributeValuesCommand.php [hunk 7]
```diff
@@ -93,94 +136,257 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        return $score;
+    }
+
+    /**
+     * Checks and removes duplicate attribute values and rii if exists.
+     *
+     * @param array $attributeValues
+     * @return void
+     */
+    protected function removeDuplicateAttributeValuesAndRii(array $attributeValues): void
```
[the remark this answers is not in the document]
Comment (pixelmatseriks) [reply Classes/Command/FixDuplicateAttributeValuesCommand.php:376]: old code, removed now, but meant tx_pxaproductmanager_relation_inheritance_index

**Pull request merged**